docs/src/apalache/fold.md

```diff
@@ -0,0 +1,213 @@
+
+### Using fold operators in Apalache
+
+The syntax of the fold operators is as follows:
+
+```
+\* @type: ( (a, b) => a, a, Set(b) ) => a
+FoldSet( operator, base, set )
+
+\* @type: ( (a, b) => a, a, Seq(b) ) => a
+FoldSeq( operator, base, seq )
+```
+
+As shown by the type signature, Apalache permits a very general form of folding, where the types of the collection elements and the type of the base element/return-type of the operator do not have to match. Again, we urge users to exercise caution when using `FoldSet` with an operator, for which the types `a` and `b` are different, as such operators cannot be commutative or associative, and therefore the result is not guaranteed to be unique and predictable. 
+
+The other component of note is `operator`, the _name_ (not definition) of some binary operator, which is available in this context. The following are examples of valid uses of folds:
+
+```
+PlusOne(p,q) == p + q + 1
+X == FoldSet( PlusOne, 0, {1,2,3} ) \* X = 9 
+
+X == LET Count(p,q) == p + 1 IN FoldSeq( Count, 0, <<1,2,3>> ) \* X = 3
+```
+
+while these next examples are considered invalid:
+
+```
+\* LAMBDAS in folds are not supported
+\* Use a local LET-IN operator instead 
+X == FoldSet( LAMBDA p,q: p + q, 0, {1,2,3} ) 
+
+\* Built-in operators cannot be called by name
+\* Define a LET-IN operator Plus(p,q) == p + q instead
+X == FoldSet( + , 0, {1,2,3} )
+```
+
+Local LET definitions can also be used as closures:
+
+```
+A(x) == LET PlusX(p,q) == p + q + x IN FoldSeq( PlusX, 0, <<1,2,3>> )
+
+X == A(1) \* X = 9
+```
+
+### Folding VS recursion
+
+While TLA+ allows users to write arbitrary recursive operators, they are, in our experience, mostly used to implement collection traversals. Consider the following implementations of a `Max` operator, which returns the largest element of a sequence:
+
+```
+CONSTANT negInf \* Max(<<>>) = -inf, but TLA+ has no literal for ±inf, so we use it as a constant
+
+RECURSIVE MaxRec(_)
+MaxRec(seq) == IF seq = <<>>
+               THEN negInf
+               ELSE LET tailMax == MaxRec(Tail(seq))
+                    IN IF tailMax > Head(seq)
+                       THEN tailMax
+                       ELSE Head(seq)
+
+MaxFold(seq) == LET Max(p,q) == IF p > q THEN p ELSE q
+                IN FoldSeq( Max, negInf, seq )
+```
+
+The first advantage of the fold implementation, we feel, is that it is much more clear and concise. It also does not require a termination condition, unlike the recursive case.
+One inherent problem of using recursive operators with a symbolic encoding, is the inability to estimate termination.
+While it may be immediately obvious to a human, that `MaxRec` terminates after no more than `Len(seq)` steps, automatic termination analysis is, in general, a rather complex form of static analysis.
```

Comment:
```suggestion
While it may be immediately obvious to a human, that `MaxRec` terminates after no more than `Len(seq)` steps, automatic termination analysis is, in general, a rather complex and incomplete form of static analysis.
```